bigframes/session/loader.py

```diff
@@ -0,0 +1,641 @@
+# limitations under the License.
+
+from __future__ import annotations
+
+import copy
+import dataclasses
+import datetime
+import itertools
+import typing
+from typing import Dict, Hashable, IO, Iterable, List, Optional, Sequence, Tuple, Union
+
+import bigframes_vendored.pandas.io.gbq as third_party_pandas_gbq
+import google.api_core.exceptions
+import google.auth.credentials
+import google.cloud.bigquery as bigquery
+import google.cloud.bigquery.table
+import google.cloud.bigquery_connection_v1
+import google.cloud.bigquery_storage_v1
+import google.cloud.functions_v2
+import google.cloud.resourcemanager_v3
+import jellyfish
+import numpy as np
+import pandas
+
+import bigframes.clients
+import bigframes.constants as constants
+import bigframes.core as core
+import bigframes.core.blocks as blocks
+import bigframes.core.compile
+import bigframes.core.guid
+import bigframes.core.pruning
+import bigframes.core.schema as schemata
+import bigframes.core.utils as utils
+
+# Even though the ibis.backends.bigquery import is unused, it's needed
+# to register new and replacement ops with the Ibis BigQuery backend.
+import bigframes.dataframe
+import bigframes.dtypes
+import bigframes.exceptions
+import bigframes.formatting_helpers as formatting_helpers
+import bigframes.session._io.bigquery as bf_io_bigquery
+import bigframes.session._io.bigquery.read_gbq_table as bf_read_gbq_table
+import bigframes.session.clients
+import bigframes.session.executor
+import bigframes.session.metrics
+import bigframes.session.planner
+import bigframes.session.temp_storage
+import bigframes.version
+
+# Avoid circular imports.
+if typing.TYPE_CHECKING:
+    import bigframes.core.indexes
+    import bigframes.dataframe as dataframe
+    import bigframes.series
+    import bigframes.session
+
+_MAX_CLUSTER_COLUMNS = 4
+
+
+def _to_index_cols(
+    index_col: Iterable[str] | str | bigframes.enums.DefaultIndexKind = (),
+) -> List[str]:
+    """Convert index_col into a list of column names."""
+    if isinstance(index_col, bigframes.enums.DefaultIndexKind):
+        index_cols: List[str] = []
+    elif isinstance(index_col, str):
+        index_cols = [index_col]
+    else:
+        index_cols = list(index_col)
+
+    return index_cols
+
+
+@dataclasses.dataclass
+class GbqDataLoader:
+    """
+    Responsible for loading data into BigFrames using temporary bigquery tables.
+
+    This loader is constrained to loading local data and queries against data sources in the same region as the storage manager.
+    """
+
+    def __init__(
+        self,
+        session: bigframes.session.Session,
+        bqclient: bigquery.Client,
+        storage_manager: bigframes.session.temp_storage.TemporaryGbqStorageManager,
+        default_index_type: bigframes.enums.DefaultIndexKind,
+        scan_index_uniqueness: bool,
+        metrics: Optional[bigframes.session.metrics.ExecutionMetrics] = None,
+    ):
+        self.bqclient = bqclient
```

Comment:
Some members are "private" while others are public. Are they really different or they can be consistent?

Reply:
made them all private